marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/main.sjs

```diff
@@ -102,22 +103,27 @@ function main(contentSequence, options, stepExecutionContext) {
           entityContext = createContextForRelatedEntityInstance(currentRelatedMapping);
         }
         const entityModel = entityModelMap[entityName];
-    
+
         for(const entityInstance of xdmp.arrayValues(instanceArray)){
           let entityContent = {};
           if(entityName == targetEntityName){
             entityContent = Object.assign(entityContent, content);
           }
           entityContent["value"] = entityInstance.value;
-          entityContent["uri"] = flowUtils.properExtensionURI(String(entityInstance.uri), outputFormat);
+          if (String(entityInstance.uri)){
+            entityContent["uri"] = flowUtils.properExtensionURI(String(entityInstance.uri), outputFormat);
+          }
+          else{
+            httpUtils.throwBadRequest(`Unable to write mapped instance for entity model '${entityName}'; cause: The uri xpath expression for the mapping evaluates to null`);
+          }
```

Comment:
This function is already too long, and this is a for loop inside of a for loop, making for a very complex function to understand. It looks like we can at least put this new chunk of logic into a function that returns a valid URI or throws an error - e.g. buildUri(entityInstance, entityName, outputFormat).